### ISSUE TITLE
Huge CPU Usage under Wayland

### ISSUE THREAD

**Issue by jorenko** [tiliado/nuvolaplayer#260]
**This issue has been fixed in the [master](https://github.com/tiliado/nuvolaplayer) git branch. Thanks for making Nuvola Player better by filing bug reports.**

The fix is available in the latest [flatpak builds](https://github.com/tiliado/nuvolaplayer/wiki/Flatpaks). Later, it will be available with the next rebuilt of packages of the [Nuvola Player 3.1 Rolling Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.1/install.html) and alsoincluded in the next bug-fix release of the [Nuvola Player 3.0 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html). To get informed when Nuvola Player 3.0.5 Stable Release is released, follow Nuvola Player on [Facebook](https://www.facebook.com/nuvolaplayer), [Google+](https://plus.google.com/110794636546911932554) or [Twitter](https://twitter.com/NuvolaPlayer); or subscribe to the [Nuvola Player Newsletter](http://eepurl.com/bLbtM1).

---

I use nuvola with google play on Arch linux. While nuvola is running, WebKitWebProcess uses about 530% of CPU on my 8-core system -- adding about 60-70% load on all 8 cores at all times.

**Comment by fenryxo** [comment]
Hello @jorenko. Thanks for taking the time to report your issue. I need more information to investigate it further:

* Which desktop environment and GTK+ theme do you use?
* Is the issue reproducible in the [WebKitGTK MiniBrowser](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines#webkitgtk-minibrowser)?
* Did it always happen or is it a result of recent package updates on your system?

**Comment by fenryxo** [comment]
Hello @jorenko. I'm closing this issue as I haven't received any response to my request for more information I posted on October 22, 2016. If the issue still persists, feel free to add missing information and reopen this ticket.

**Comment by excieve** [comment]
I'm experiencing this too, at least with Google Play Music. Whenever there's any kind of activity in the window the CPU usage maxes out on all available cores. In case of no activity or window being closed (player running in the background) there's no such problem.

> Which desktop environment and GTK+ theme do you use?
Gnome 3.22.2 with "Arc" theme ("Darker" variant) running on Wayland (didn't try X11).
> Is the issue reproducible in the WebKitGTK MiniBrowser?
No, CPU usage is normal with it.
> Did it always happen or is it a result of recent package updates on your system?
I haven't been using Nuvola for some time but haven't noticed this before. Currently on version 3.1.1~201701141634.57169db.

**Comment by fenryxo** [comment]
Hello @excieve. Thanks for reporting your issue.

>> Which desktop environment and GTK+ theme do you use?

> Gnome 3.22.2 with "Arc" theme ("Darker" variant) running on Wayland (didn't try X11).

Is the issue reproducible if you switch to the Adwaita theme?

>> Is the issue reproducible in the WebKitGTK MiniBrowser?

> No, CPU usage is normal with it.

>> Did it always happen or is it a result of recent package updates on your system?

> I haven't been using Nuvola for some time but haven't noticed this before. Currently on version 3.1.1~201701141634.57169db.

It might be related to the issue #279.

**Comment by excieve** [comment]
@fenryxo Same behaviour on Adwaita. As for #279 — nothing freezes for me, just huge CPU usage. Also probably unrelated but there's lots of channel debug messages spam in the system log such as:
```
eu.tiliado.Nuvola.desktop[2673]: Runner: [WARNING Δ502295us DioriteGlib] MessageChannel.vala:181: on incoming request 174 /nuvola/core/config-get-value::prw,tuple,c80080cc99c956f171141310445384a6c062d1ffc9ab46e1beb91d021ac0af85
```

**Comment by fenryxo** [comment]
Hello @excieve. Could you test the latest build of both Diorite library & Nuvola Player 3.1? Thanks.

**Comment by excieve** [comment]
@fenryxo Unfortunately CPU usage is still huge with nuvolaplayer3-core-3.1.1~201702051432.69be32a-1.fc25.x86_64 and diorite0.3-0.3.1~201702051435.8139a48-1.fc25.x86_64. Excessive debug messages are gone though.

**Comment by fenryxo** [comment]
Could you look into which process is responsible for that? (There are at least 5 processes: Nuvola Master, Nuvola App Runner, WebKit WebProcess, WebKit DatabaseProcess and WebKit NetworkProcess).

Also, could you check whether the issue occurs also with [Flatpak builds](https://github.com/tiliado/nuvolaplayer/wiki/Flatpaks)?

**Comment by excieve** [comment]
This is the WebKitWebProcess and its many threads. To add more detail the CPU usage soars when playing something or moving the pointer over various elements of the page (but not over the empty spaces). Pausing brings the CPU usage back to normal.

I'll try a flatpak build a bit later today.

**Comment by excieve** [comment]
@fenryxo I have installed Flatpak following the instructions in your link but I have issues running it. It won't launch Google Play Music app probably due to requiring a premium/patron account, which I have but "Activate Nuvola Apps" button just shows a message telling "This build of Nuvola Apps is not authorized to use the Tiliado API."

However, I logged into an X11 session and CPU usage is absolutely normal with the native package there. So apparently this is a Wayland-related issue. Still, "MiniBrowser" doesn't exhibit such behaviour in Wayland when launched with XWayland backend like this:
GDK_BACKEND=x11 /usr/libexec/webkit2gtk-4.0/MiniBrowser

**Comment by fenryxo** [comment]
> but "Activate Nuvola Apps" button just shows a message telling "This build of Nuvola Apps is not authorized to use the Tiliado API."

This should be fixed now.

> However, I logged into an X11 session and CPU usage is absolutely normal with the native package there. So apparently this is a Wayland-related issue. Still, "MiniBrowser" doesn't exhibit such behaviour in Wayland when launched with XWayland backend like this:
GDK_BACKEND=x11 /usr/libexec/webkit2gtk-4.0/MiniBrowser

Does the issue occur also with other Nuvola apps (e.g. Deezer) or is it specific to Google Play Music?

**Comment by excieve** [comment]
> This should be fixed now.

Now it apparently hangs at "Authorization procedure in progress..." or at least nothing visible happens for several minutes before I hit "Cancel".

> Does the issue occur also with other Nuvola apps (e.g. Deezer) or is it specific to Google Play Music?

Yes, exactly the same issue with Deezer — normal load when nothing happens on the page or window closed, high load when hovering over UI elements or playing music with window open.

**Comment by fenryxo** [comment]
> Now it apparently hangs at "Authorization procedure in progress..." or at least nothing visible happens for several minutes before I hit "Cancel".

A web page should be opened when "Authorization procedure in progress..." is shown. It seems as if Freedesktop Desktop Portal wasn't installed. Are the `xdg-desktop-portal` and `xdg-desktop-portal-gtk` installed? (Related #293)

**Comment by excieve** [comment]
Right, installing `xdg-desktop-portal` and `xdg-desktop-portal-gtk` did it, thanks! Behaviour with Flatpak build is somewhat different. There's still same CPU load when hovering over various UI elements on the page but playing a song with window open results in about 40-50% load for the WebKitWebProcess. This is much better but still pretty high I think.

**Comment by fenryxo** [comment]
Could you `flatpak update` and test the latest build? I've disabled an old workaround that might no be necessary with a recent WebKitGTK and may reduce rendering performance.

**Comment by excieve** [comment]
@fenryxo This fixed it apparently. WebKitWebProcess in now under 4% CPU when playing with window open and no huge spikes on hovering over UI elements. Thank you for looking into it!

**Comment by fenryxo** [comment]
Great :-) For the reference, the old issue was "Scrolling lag or slow repaint" #24 and was trigerred by a bug in certain graphics drivers. Since Nuvola is moving to flatpak, I can finally remove a lot of legacy code and problematic workarounds.

**Comment by fenryxo** [comment]
**This issue has been fixed in the [master](https://github.com/tiliado/nuvolaplayer) git branch. Thanks for making Nuvola Player better by filing bug reports.**

The fix is available in the latest [flatpak builds](https://github.com/tiliado/nuvolaplayer/wiki/Flatpaks). Later, it will be available with the next rebuilt of packages of the [Nuvola Player 3.1 Rolling Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.1/install.html) and also included in the next bug-fix release of the [Nuvola Player 3.0 Stable Releases](https://tiliado.github.io/nuvolaplayer/documentation/3.0/install.html). To get informed when Nuvola Player 3.0.5 Stable Release is released, follow Nuvola Player on [Facebook](https://www.facebook.com/nuvolaplayer), [Google+](https://plus.google.com/110794636546911932554) or [Twitter](https://twitter.com/NuvolaPlayer); or subscribe to the [Nuvola Player Newsletter](http://eepurl.com/bLbtM1).

**Comment by fenryxo** [comment]
[Nuvola Player 3.0.5 and Diorite 0.2.2 has been released.](https://medium.com/nuvola-news/nuvola-player-3-0-5-and-diorite-0-2-2-released-c170dfb53130)
